### ISSUE TITLE
Java OpenJDK 7 Root Certs and Scala SBT 0.13.13 - DevOps 2.0 Toolkit Pg. 72

### ISSUE THREAD

**Issue by TorrentialFire** [vfarcic/books-ms#60]
It looks like there might be a new issue related to:
[Dockerfile.test error #33](https://github.com/vfarcic/books-ms/issues/33)

I am working through the DevOps 2.0 Toolkit book, and I have encountered the following issue on page 72 while running the following command:

```
docker build \
-f Dockerfile.test \
-t 10.100.198.200:5000/books-ms-tests \
.
```
I have made sure to clone the github repo into the cd virtual machine and I am ssh'd in. My cd virtual machine was built with vagrant 2.2.5 and I am running VirtualBox 6.0.10 on a Windows 10 Pro (Version 1903, Build 18362.295).

The container is built up to setting up openJDK-7. The following error occurs during setup (a few lines are included around the stacktrace for context):

```
update-alternatives: using /usr/lib/jvm/java-7-openjdk-amd64/jre/bin/servertool to provide /usr/bin/servertool (servertool) in auto mode
update-alternatives: using /usr/lib/jvm/java-7-openjdk-amd64/jre/bin/tnameserv to provide /usr/bin/tnameserv (tnameserv) in auto mode
update-alternatives: using /usr/lib/jvm/java-7-openjdk-amd64/jre/lib/jexec to provide /usr/bin/jexec (jexec) in auto mode
Setting up ca-certificates-java (20140324) ...
Adding debian:Hellenic_Academic_and_Research_Institutions_RootCA_2011.pem
Exception in thread "main" java.lang.NoClassDefFoundError: sun/security/ec/ECParameters
        at sun.security.pkcs11.SunPKCS11$P11Service.newInstance0(SunPKCS11.java:1038)
        at sun.security.pkcs11.SunPKCS11$P11Service.newInstance(SunPKCS11.java:980)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
        at java.security.Security.getImpl(Security.java:694)
        at java.security.AlgorithmParameters.getInstance(AlgorithmParameters.java:146)
        at sun.security.x509.AlgorithmId.decodeParams(AlgorithmId.java:121)
        at sun.security.x509.AlgorithmId.<init>(AlgorithmId.java:114)
        at sun.security.x509.AlgorithmId.parse(AlgorithmId.java:372)
        at sun.security.x509.X509Key.parse(X509Key.java:167)
        at sun.security.x509.CertificateX509Key.<init>(CertificateX509Key.java:75)
        at sun.security.x509.X509CertInfo.parse(X509CertInfo.java:699)
        at sun.security.x509.X509CertInfo.<init>(X509CertInfo.java:166)
        at sun.security.x509.X509CertImpl.parse(X509CertImpl.java:1806)
        at sun.security.x509.X509CertImpl.<init>(X509CertImpl.java:197)
        at sun.security.provider.X509Factory.engineGenerateCertificate(X509Factory.java:98)
        at java.security.cert.CertificateFactory.generateCertificate(CertificateFactory.java:339)
        at org.debian.security.KeyStoreHandler.loadCertificate(KeyStoreHandler.java:138)
        at org.debian.security.KeyStoreHandler.addAlias(KeyStoreHandler.java:94)
        at org.debian.security.UpdateCertificates.parseLine(UpdateCertificates.java:93)
        at org.debian.security.UpdateCertificates.processChanges(UpdateCertificates.java:76)
        at org.debian.security.UpdateCertificates.main(UpdateCertificates.java:53)
Caused by: java.lang.ClassNotFoundException: sun.security.ec.ECParameters
        at java.net.URLClassLoader$1.run(URLClassLoader.java:360)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:349)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:348)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:430)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
        ... 22 more
done.
Setting up sbt (0.13.13) ...
Setting up libgtk-3-common (3.14.5-1+deb8u1) ...
Setting up libgtk-3-0:amd64 (3.14.5-1+deb8u1) ...
Setting up libgtk-3-bin (3.14.5-1+deb8u1) ...
Setting up adwaita-icon-theme (3.14.0-2) ...
```

It looks like java encounters an issue setting up a root certificate but continues anyways.

The setup fails at the following step:

```
+ bower@1.8.8
added 315 packages from 218 contributors in 20.495s
 ---> 4e05e9e3d0f2
Removing intermediate container f3b6aa245110
Step 10/29 : RUN mkdir /source
 ---> Running in 32ed6a22685c
 ---> 49da56c908db
Removing intermediate container 32ed6a22685c
Step 11/29 : RUN mkdir -p /data/db
 ---> Running in 7ef73f0d653b
 ---> 15bae930148b
Removing intermediate container 7ef73f0d653b
Step 12/29 : ADD project /source/project
 ---> 398e4fc4632f
Removing intermediate container d7db3d88fbc9
Step 13/29 : ADD build.sbt /source/build.sbt
 ---> 3457b6010edc
Removing intermediate container eb9c92e7e77b
Step 14/29 : ADD client/bower.json /source/client/bower.json
 ---> da6ebd8df27c
Removing intermediate container e4e38e7cd0d7
Step 15/29 : ADD client/gulpfile.js /source/client/gulpfile.js
 ---> f9724b397592
Removing intermediate container 22bc75633356
Step 16/29 : ADD client/package.json /source/client/package.json
 ---> f71302345041
Removing intermediate container 326c78da0719
Step 17/29 : ADD client/wct.conf.js /source/client/wct.conf.js
 ---> 6da18cb4d4f2
Removing intermediate container 1625f8f7549d
Step 18/29 : ADD client/test.html /source/client/test.html
 ---> 7d7d66dc869d
Removing intermediate container 027699e742c3
Step 19/29 : ADD run_tests.sh /source/run_tests.sh
 ---> 0bf4054f3a78
Removing intermediate container 7c0af52b6394
Step 20/29 : RUN cd /source && sbt update
 ---> Running in c4b4ce69ca9c
Getting org.scala-sbt sbt 0.13.13 ...

:: problems summary ::
:::: WARNINGS
                module not found: org.scala-sbt#sbt;0.13.13

        ==== local: tried

          /root/.ivy2/local/org.scala-sbt/sbt/0.13.13/ivys/ivy.xml

          -- artifact org.scala-sbt#sbt;0.13.13!sbt.jar:

          /root/.ivy2/local/org.scala-sbt/sbt/0.13.13/jars/sbt.jar

        ==== Maven Central: tried

          https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.13/sbt-0.13.13.pom

          -- artifact org.scala-sbt#sbt;0.13.13!sbt.jar:

          https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.13/sbt-0.13.13.jar

        ==== typesafe-ivy-releases: tried

          https://repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/sbt/0.13.13/ivys/ivy.xml

        ==== sbt-ivy-snapshots: tried

          https://repo.scala-sbt.org/scalasbt/ivy-snapshots/org.scala-sbt/sbt/0.13.13/ivys/ivy.xml

                ::::::::::::::::::::::::::::::::::::::::::::::

                ::          UNRESOLVED DEPENDENCIES         ::

                ::::::::::::::::::::::::::::::::::::::::::::::

                :: org.scala-sbt#sbt;0.13.13: not found

                ::::::::::::::::::::::::::::::::::::::::::::::


:::: ERRORS
        Server access Error: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty url=https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.13/sbt-0.13.13.pom

        Server access Error: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty url=https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.13/sbt-0.13.13.jar

        Server access Error: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty url=https://repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/sbt/0.13.13/ivys/ivy.xml

        Server access Error: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty url=https://repo.scala-sbt.org/scalasbt/ivy-snapshots/org.scala-sbt/sbt/0.13.13/ivys/ivy.xml


:: USE VERBOSE OR DEBUG MESSAGE LEVEL FOR MORE DETAILS
unresolved dependency: org.scala-sbt#sbt;0.13.13: not found
Error during sbt execution: Error retrieving required libraries
  (see /root/.sbt/boot/update.log for complete log)
Error: Could not retrieve sbt 0.13.13
The command '/bin/sh -c cd /source && sbt update' returned a non-zero code: 1
```
It seems like someone encountered a similar issue (with java root certs and scala SBT) and posted in the SBT repo:
[https://github.com/sbt/sbt/issues/2295](https://github.com/sbt/sbt/issues/2295)

Successive attempts to call the docker build command again result in the same error.

Reading from the issued linked above and looking in Dockerfile.test, I modified the RUN apt-get commands, adding a `RUN update-ca-certificates -f -v` after:

```
RUN apt-get update && \
    apt-get install -y --fix-missing~~~
    apt-get -y --force~~~
    apt-get clean && \
    rm -rf ~

RUN update-ca-certificates -f -v
```
This gets me further, but there is a problem while running `/etc/ca-certificates/update.d`.

```
Amazon_Root_CA_3.pem => 8cb5ee0f.0
Amazon_Root_CA_3.pem => 7a7c655d.0
AC_RAIZ_FNMT-RCM.pem => cd8c0d63.0
AC_RAIZ_FNMT-RCM.pem => b936d1c6.0
Hellenic_Academic_and_Research_Institutions_ECC_RootCA_2015.pem => 7719f463.0
Hellenic_Academic_and_Research_Institutions_ECC_RootCA_2015.pem => 9479c8c3.0
152 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d....
Adding debian:ACCVRAIZ1.pem
Adding debian:AC_RAIZ_FNMT-RCM.pem
Adding debian:AC_Raíz_Certicámara_S.A..pem
Adding debian:Actalis_Authentication_Root_CA.pem
Adding debian:AddTrust_External_Root.pem
Adding debian:AddTrust_Low-Value_Services_Root.pem
Adding debian:AffirmTrust_Commercial.pem
Adding debian:AffirmTrust_Networking.pem
Adding debian:AffirmTrust_Premium.pem
Exception in thread "main" java.lang.NoClassDefFoundError: sun/security/ec/ECParameters
        at sun.security.pkcs11.SunPKCS11$P11Service.newInstance0(SunPKCS11.java:1038)
        at sun.security.pkcs11.SunPKCS11$P11Service.newInstance(SunPKCS11.java:980)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
        at java.security.Security.getImpl(Security.java:694)
        at java.security.AlgorithmParameters.getInstance(AlgorithmParameters.java:146)
        at sun.security.x509.AlgorithmId.decodeParams(AlgorithmId.java:121)
        at sun.security.x509.AlgorithmId.<init>(AlgorithmId.java:114)
        at sun.security.x509.AlgorithmId.parse(AlgorithmId.java:372)
        at sun.security.x509.X509Key.parse(X509Key.java:167)
        at sun.security.x509.CertificateX509Key.<init>(CertificateX509Key.java:75)
        at sun.security.x509.X509CertInfo.parse(X509CertInfo.java:699)
        at sun.security.x509.X509CertInfo.<init>(X509CertInfo.java:166)
        at sun.security.x509.X509CertImpl.parse(X509CertImpl.java:1806)
        at sun.security.x509.X509CertImpl.<init>(X509CertImpl.java:197)
        at sun.security.provider.X509Factory.engineGenerateCertificate(X509Factory.java:98)
        at java.security.cert.CertificateFactory.generateCertificate(CertificateFactory.java:339)
        at org.debian.security.KeyStoreHandler.loadCertificate(KeyStoreHandler.java:138)
        at org.debian.security.KeyStoreHandler.addAlias(KeyStoreHandler.java:94)
        at org.debian.security.UpdateCertificates.parseLine(UpdateCertificates.java:93)
        at org.debian.security.UpdateCertificates.processChanges(UpdateCertificates.java:76)
        at org.debian.security.UpdateCertificates.main(UpdateCertificates.java:53)
Caused by: java.lang.ClassNotFoundException: sun.security.ec.ECParameters
        at java.net.URLClassLoader$1.run(URLClassLoader.java:360)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:349)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:348)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:430)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
        ... 22 more
E: /etc/ca-certificates/update.d/jks-keystore exited with code 1.
done.
```
OpenJDK 7u231 has some issues in this case, I guess. I'm not sure how to proceed.

**Comment by vfarcic** [comment]
The tech and the processes used in 2.0 are tool old. Some things are outdated, some tools are not used any more, and some principles are today obsolete. I've been keeping the book up-to-date for a long time but over a year ago I decided to stop and remove it from sales (at least from those vendors I do control). The main reason for removal was not because of the difficulties in maintaining it but rather because I do not think it represents what I believe anymore.

All in all, the book and the examples in it are obsolete/deprecated. They're not maintained anymore.

Since 2.0, I wrote 5 more books (6th is coming soon) and they are all maintained and up-to-date. I'd be more than happy to give you any of those instead. Just let me know which one you'd like and send me an email address I can use to send you a free coupon. You can find the list of books in https://www.devopstoolkitseries.com.

P.S. Please let me know where you got the book from so that I can notify that vendor to stop selling it.

**Comment by TorrentialFire** [comment]
I will send you an email for us to continue this discussion privately. I appreciate your quick response!

When I posted this issue last night, I did read some of the recent issues and replies for this book here on git and recognized that you would probably tell me the same thing (hopefully you have a pastebin or text file saved somewhere with the general response).

Might I recommend adding a disclaimer to the readme for the repo, and in some of the major files maybe (especially files the book recommends to take a look at) that informs the reader that this project isn't being actively maintained and they should proceed at their own risk? A commit message along with that explaining the state of the repository would help reader realize more quickly that they are using unsupported files and there will be challenges. It's probably for the best that you leave the repo up, it would be even worse to get reviews/messages from disgruntled readers who are angry that they can't even clone the repo.
